src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/References/ReferenceCleanupService.ReferenceHandlers.cs

```diff
@@ -0,0 +1,185 @@
+
+                var item = projectItems
+                    .FirstOrDefault(c => c.UnevaluatedInclude == referenceUpdate.ReferenceInfo.ItemSpecification);
+
+                if (item != null)
+                {
+                    string newValue = referenceUpdate.Action == ProjectSystemUpdateAction.TreatAsUsed ? bool.TrueString : bool.FalseString;
+
+                    await item.Metadata.SetPropertyValueAsync(ProjectReference.TreatAsUsedProperty, newValue, null);
+
+                    wasUpdated = true;
+                }
+
+                return wasUpdated;
+            }
+        }
+
+        private class ProjectReferenceHandler : AbstractReferenceHandler
+        {
+            internal ProjectReferenceHandler() :
+                base(ProjectSystemReferenceType.Project)
+            { }
+
+            protected override Task RemoveReferenceAsync(ConfiguredProjectServices services,
+                ProjectSystemReferenceInfo referencesInfo)
+            {
+                Assumes.Present(services.ProjectReferences);
+
+                return services.ProjectReferences.RemoveAsync(referencesInfo.ItemSpecification);
+            }
+
+            protected override async Task<IEnumerable<IProjectItem>> GetUnresolvedReferencesAsync(ConfiguredProjectServices services)
+            {
+                Assumes.Present(services.ProjectReferences);
+
+                return (await services.ProjectReferences.GetUnresolvedReferencesAsync()).Cast<IProjectItem>();
+            }
+        }
+
+        private class PackageReferenceHandler : AbstractReferenceHandler
+        {
+            internal PackageReferenceHandler() :
+                base(ProjectSystemReferenceType.Package)
+            { }
+
+            protected override Task RemoveReferenceAsync(ConfiguredProjectServices services,
+                ProjectSystemReferenceInfo referencesInfo)
+            {
+                Assumes.Present(services.PackageReferences);
+
+                return services.PackageReferences.RemoveAsync(referencesInfo.ItemSpecification);
+            }
+
+            protected override async Task<IEnumerable<IProjectItem>> GetUnresolvedReferencesAsync(ConfiguredProjectServices services)
+            {
+                Assumes.Present(services.PackageReferences);
+
+                return (await services.PackageReferences.GetUnresolvedReferencesAsync()).Cast<IProjectItem>();
+            }
+        }
+
+        private class AssemblyReferenceHandler : AbstractReferenceHandler
+        {
+            internal AssemblyReferenceHandler() :
+                base(ProjectSystemReferenceType.Assembly)
+            { }
+
+            protected override Task RemoveReferenceAsync(ConfiguredProjectServices services,
+                ProjectSystemReferenceInfo referencesInfo)
+            {
+                Assumes.Present(services.AssemblyReferences);
+
+                var assemblyName = new AssemblyName(referencesInfo.ItemSpecification);
```

Comment:
@davkean 
UnevaluatedInclude will break AssemblyName ? Should we use full path instead ?

Reply:
ItemSpecification is not always an "assembly name":

``` XML
<Reference Include="C:\Foo\System.dll"/>
```

You'll need to handle both cases.

Reply:
I added code to check when ItemSpecification is path or name.